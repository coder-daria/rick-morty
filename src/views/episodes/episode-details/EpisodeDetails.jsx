import React, { useMemo } from 'react';
import { instanceOf, shape, string } from 'prop-types';
import { Col } from 'antd';

import CharacterModel from '../model/EpisodeModel';

import { StyledRow } from './EpisodeDetails.styles';

const NO_DATA_PLACEHOLDER = '-';

function renderCharactersNames(characters) {
  return characters.map(({ name, index }) => (
    <div key={`${name + index}`}>{name}</div>
  ));
}

function EpisodeDetails({ episode }) {
  const characterDetails = useMemo(() => CharacterModel(episode), [episode]);

  return (
    <div>
      {Object.keys(characterDetails).map(item => (
        <StyledRow key={item}>
          <Col span={12}>{item}</Col>
          <Col span={12}>
            {item === 'characters'
              ? renderCharactersNames(characterDetails[item])
              : characterDetails[item] || NO_DATA_PLACEHOLDER}
          </Col>
        </StyledRow>
      ))}
    </div>
  );
}

EpisodeDetails.propTypes = {
  episode: shape({
    air_date: string,
    characters: instanceOf(Array),
    created: string,
    episode: string,
    id: string,
    name: string,
  }).isRequired,
};

export default EpisodeDetails;
