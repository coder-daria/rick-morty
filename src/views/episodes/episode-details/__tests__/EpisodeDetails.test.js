import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import {
  EPISODE_DETAILS_MOCKED_DATA,
  EPISODE_NOT_FORMATED_MOCKED_DETAILS,
} from '../../__mocks__/episode-mock-data';

import EpisodeDetails from '../EpisodeDetails';

const setup = () =>
  shallow(
    <EpisodeDetails
      episode={{
        ...EPISODE_DETAILS_MOCKED_DATA,
        ...EPISODE_NOT_FORMATED_MOCKED_DETAILS,
      }}
    />,
  );

describe('<EpisodeDetails />', () => {
  it('should match the snapshot', () => {
    const wrapper = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
