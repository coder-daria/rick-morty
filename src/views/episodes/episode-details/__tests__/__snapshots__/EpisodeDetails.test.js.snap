// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EpisodeDetails /> should match the snapshot 1`] = `
<div>
  <Styled(Row)
    key="name"
  >
    <Col
      span={12}
    >
      name
    </Col>
    <Col
      span={12}
    >
      Sunny day
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="episode"
  >
    <Col
      span={12}
    >
      episode
    </Col>
    <Col
      span={12}
    >
      0E12
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="characters"
  >
    <Col
      span={12}
    >
      characters
    </Col>
    <Col
      span={12}
    >
      <div
        key="Tomundefined"
      >
        Tom
      </div>
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="air date"
  >
    <Col
      span={12}
    >
      air date
    </Col>
    <Col
      span={12}
    >
      11/25/2020, 4:59:08 PM
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="created"
  >
    <Col
      span={12}
    >
      created
    </Col>
    <Col
      span={12}
    >
      11/25/2019, 4:59:08 PM
    </Col>
  </Styled(Row)>
</div>
`;
